lib/vrl/stdlib/src/zip.rs

```diff
@@ -0,0 +1,170 @@
+    for (key, value) in keys.iter().zip(items.into_iter()) {
+        let key = key.try_bytes_utf8_lossy()?;
+        match map.entry(key.into_owned()) {
+            Entry::Vacant(entry) => {
+                entry.insert(value);
+            }
+            Entry::Occupied(mut entry) => {
+                // always replace previous entry
+                let existing = entry.get_mut();
+                *existing = value;
+            }
+        }
+    }
+    Ok(Value::Object(map))
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct Zip;
+
+impl Function for Zip {
+    fn identifier(&self) -> &'static str {
+        "zip"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "keys",
+                kind: kind::ARRAY,
+                required: true,
+            },
+            Parameter {
+                keyword: "items",
+                kind: kind::ARRAY,
+                required: true,
+            },
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "zip arrays into an object (first must be strings)",
+                source: r#"zip(["a", "b"], [1, 2])"#,
+                result: Ok("{\"a\": 1, \"b\": 2}"),
+            },
+            Example {
+                title: "zip arrays of strings",
+                source: r#"zip(["a", "b"], ["c", "d"])"#,
+                result: Ok("{\"a\": \"c\", \"b\": \"d\"}"),
+            },
+            Example {
+                title: "zip duplicate keys (last one wins)",
+                source: r#"zip(["a", "a"], ["c", "d"])"#,
+                result: Ok("{\"a\": \"d\"}"),
+            },
+            Example {
+                title: "odd length arrays iterate to shortest length",
+                source: r#"zip(["a", "b"], [1])"#,
+                result: Ok("{\"a\": 1}"),
+            },
+        ]
+    }
+
+    fn compile(
+        &self,
+        _state: &state::TypeState,
+        _ctx: &mut FunctionCompileContext,
+        arguments: ArgumentList,
+    ) -> Compiled {
+        let keys = arguments.required("keys");
+        let items = arguments.required("items");
+
+        Ok(ZipFn { keys, items }.as_expr())
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ZipFn {
+    keys: Box<dyn Expression>,
+    items: Box<dyn Expression>,
+}
+
+impl FunctionExpression for ZipFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let keys = self.keys.resolve(ctx)?;
+        let items = self.items.resolve(ctx)?;
+
+        zip(keys, items)
+    }
+
+    fn type_def(&self, state: &state::TypeState) -> TypeDef {
+        let keys = self.keys.type_def(state).restrict_array();
+        let items = self.items.type_def(state).restrict_array();
+
+        let _ = keys.as_array().expect("must be an array");
+        let _ = items.as_array().expect("must be an array");
+
+        TypeDef::object(Collection::any()).infallible()
+    }
```

Comment:
I don't fully grok the type_def functionality in the vrl yet but, since the BTreeMap keys are being dynamically defined by the fn arguments I don't know if a more descriptive [BTreeMap definition is possible like in `parse_aws_alb_log.rs`](https://github.com/vectordotdev/vector/blob/master/lib/vrl/stdlib/src/parse_aws_alb_log.rs#L75-L84)

also the `.infallible()` use may be incorrect 😅